Graceful Shutdown Errors

I try to stop the pipeline but my program crashes

```go
    if err := i.pipeline.Stop(); err != nil {
        logger.Errorf("failed to stop pipeline: %s", err)
        os.Exit(1)
    }
```

```
panic: close of closed channel

goroutine 240 [running]:
github.com/blinklabs-io/adder/pipeline.(*Pipeline).Stop(0xc00049eaf0)
        go/pkg/mod/github.com/blinklabs-io/adder@v0.25.1/pipeline/pipeline.go:110 +0x25
```

![image](https://github.com/user-attachments/assets/bc38074b-224a-44d5-9b3d-6fd4bd498597)

my use case is that I detect something is out of sync. So I want to restart the indexer instance to start back from an old intercept point.

--- a/filter/chainsync/chainsync.go
+++ b/filter/chainsync/chainsync.go
@@ -17,6 +17,7 @@ package chainsync
 import (
 	"encoding/hex"
 	"strings"
+	"sync"
 
 	"github.com/blinklabs-io/adder/event"
 	"github.com/blinklabs-io/adder/plugin"
@@ -26,9 +27,11 @@ import (
 )
 
 type ChainSync struct {
-	errorChan               chan<- error
+	errorChan               chan error
 	inputChan               chan event.Event
 	outputChan              chan event.Event
+	doneChan                chan struct{}
+	stopOnce                sync.Once
 	logger                  plugin.Logger
 	filterAddresses         []string
 	filterAssetFingerprints []string
@@ -38,10 +41,7 @@ type ChainSync struct {
 
 // New returns a new ChainSync object with the specified options applied
 func New(options ...ChainSyncOptionFunc) *ChainSync {
-	c := &ChainSync{
-		inputChan:  make(chan event.Event, 10),
-		outputChan: make(chan event.Event, 10),
-	}
+	c := &ChainSync{}
 	for _, option := range options {
 		option(c)
 	}
@@ -50,278 +50,292 @@ func New(options ...ChainSyncOptionFunc) *ChainSync {
 
 // Start the chain sync filter
 func (c *ChainSync) Start() error {
+	c.errorChan = make(chan error)
+	c.inputChan = make(chan event.Event, 10)
+	c.outputChan = make(chan event.Event, 10)
+	c.doneChan = make(chan struct{})
+	c.stopOnce = sync.Once{}
 	go func() {
 		// TODO: pre-process filter params to be more useful for direct comparison (#336)
 		for {
-			evt, ok := <-c.inputChan
-			// Channel has been closed, which means we're shutting down
-			if !ok {
+			select {
+			case <-c.doneChan:
 				return
-			}
-			switch v := evt.Payload.(type) {
-			case event.BlockEvent:
-				// Check pool filter
-				if len(c.filterPoolIds) > 0 {
-					filterMatched := false
-					for _, filterPoolId := range c.filterPoolIds {
-						isPoolBech32 := strings.HasPrefix(filterPoolId, "pool")
-						foundMatch := false
-						be := evt.Payload.(event.BlockEvent)
-						if be.IssuerVkey == filterPoolId {
-							foundMatch = true
-						} else if isPoolBech32 {
-							issuerBytes, err := hex.DecodeString(be.IssuerVkey)
-							if err != nil {
-								// eat this error... nom nom nom
-								continue
-							}
-							// lifted from gouroboros/ledger
-							convData, err := bech32.ConvertBits(issuerBytes, 8, 5, true)
-							if err != nil {
-								continue
-							}
-							encoded, err := bech32.Encode("pool", convData)
-							if err != nil {
-								continue
-							}
-							if encoded == filterPoolId {
-								foundMatch = true
-							}
-						}
-						if foundMatch {
-							filterMatched = true
-							break
-						}
-					}
-					// Skip the event if none of the filter values matched
-					if !filterMatched {
-						continue
-					}
+			case evt, ok := <-c.inputChan:
+				// Channel has been closed, which means we're shutting down
+				if !ok {
+					return
 				}
-			case event.TransactionEvent:
-				// Check address filter
-				if len(c.filterAddresses) > 0 {
-					filterMatched := false
-					for _, filterAddress := range c.filterAddresses {
-						isStakeAddress := strings.HasPrefix(filterAddress, "stake")
-						foundMatch := false
-						// Include resolved inputs as outputs for matching
-						allOutputs := append(v.Outputs, v.ResolvedInputs...)
-						for _, output := range allOutputs {
-							if output.Address().String() == filterAddress {
+				switch v := evt.Payload.(type) {
+				case event.BlockEvent:
+					// Check pool filter
+					if len(c.filterPoolIds) > 0 {
+						filterMatched := false
+						for _, filterPoolId := range c.filterPoolIds {
+							isPoolBech32 := strings.HasPrefix(filterPoolId, "pool")
+							foundMatch := false
+							be := evt.Payload.(event.BlockEvent)
+							if be.IssuerVkey == filterPoolId {
 								foundMatch = true
+							} else if isPoolBech32 {
+								issuerBytes, err := hex.DecodeString(be.IssuerVkey)
+								if err != nil {
+									// eat this error... nom nom nom
+									continue
+								}
+								// lifted from gouroboros/ledger
+								convData, err := bech32.ConvertBits(issuerBytes, 8, 5, true)
+								if err != nil {
+									continue
+								}
+								encoded, err := bech32.Encode("pool", convData)
+								if err != nil {
+									continue
+								}
+								if encoded == filterPoolId {
+									foundMatch = true
+								}
+							}
+							if foundMatch {
+								filterMatched = true
 								break
 							}
-							if isStakeAddress {
-								stakeAddr := output.Address().StakeAddress()
-								if stakeAddr == nil {
-									continue
-								}
-								if stakeAddr.String() == filterAddress {
+						}
+						// Skip the event if none of the filter values matched
+						if !filterMatched {
+							continue
+						}
+					}
+				case event.TransactionEvent:
+					// Check address filter
+					if len(c.filterAddresses) > 0 {
+						filterMatched := false
+						for _, filterAddress := range c.filterAddresses {
+							isStakeAddress := strings.HasPrefix(filterAddress, "stake")
+							foundMatch := false
+							// Include resolved inputs as outputs for matching
+							allOutputs := append(v.Outputs, v.ResolvedInputs...)
+							for _, output := range allOutputs {
+								if output.Address().String() == filterAddress {
 									foundMatch = true
 									break
 								}
-							}
-						}
-
-						if !foundMatch && isStakeAddress {
-							for _, certificate := range v.Certificates {
-								var credBytes []byte
-								switch cert := certificate.(type) {
-								case *common.StakeDelegationCertificate:
-									hash := cert.StakeCredential.Hash()
-									credBytes = hash[:]
-								case *common.StakeDeregistrationCertificate:
-									hash := cert.StakeCredential.Hash()
-									credBytes = hash[:]
-								default:
-									continue
-								}
-
-								convData, err := bech32.ConvertBits(credBytes, 8, 5, true)
-								if err != nil {
-									continue
-								}
-								encoded, err := bech32.Encode("stake", convData)
-								if err != nil {
-									continue
-								}
-								if encoded == filterAddress {
-									foundMatch = true
-									break
-								}
-							}
-						}
-
-						if foundMatch {
-							filterMatched = true
-							break
-						}
-					}
-					// Skip the event if none of the filter values matched
-					if !filterMatched {
-						continue
-					}
-				}
-				// Check policy ID filter
-				if len(c.filterPolicyIds) > 0 {
-					filterMatched := false
-					for _, filterPolicyId := range c.filterPolicyIds {
-						foundMatch := false
-						// Include resolved inputs as outputs for matching
-						allOutputs := append(v.Outputs, v.ResolvedInputs...)
-						for _, output := range allOutputs {
-							if output.Assets() != nil {
-								for _, policyId := range output.Assets().Policies() {
-									if policyId.String() == filterPolicyId {
+								if isStakeAddress {
+									stakeAddr := output.Address().StakeAddress()
+									if stakeAddr == nil {
+										continue
+									}
+									if stakeAddr.String() == filterAddress {
 										foundMatch = true
 										break
 									}
 								}
 							}
+
+							if !foundMatch && isStakeAddress {
+								for _, certificate := range v.Certificates {
+									var credBytes []byte
+									switch cert := certificate.(type) {
+									case *common.StakeDelegationCertificate:
+										hash := cert.StakeCredential.Hash()
+										credBytes = hash[:]
+									case *common.StakeDeregistrationCertificate:
+										hash := cert.StakeCredential.Hash()
+										credBytes = hash[:]
+									default:
+										continue
+									}
+
+									convData, err := bech32.ConvertBits(credBytes, 8, 5, true)
+									if err != nil {
+										continue
+									}
+									encoded, err := bech32.Encode("stake", convData)
+									if err != nil {
+										continue
+									}
+									if encoded == filterAddress {
+										foundMatch = true
+										break
+									}
+								}
+							}
+
 							if foundMatch {
+								filterMatched = true
 								break
 							}
 						}
-						if foundMatch {
-							filterMatched = true
-							break
+						// Skip the event if none of the filter values matched
+						if !filterMatched {
+							continue
 						}
 					}
-					// Skip the event if none of the filter values matched
-					if !filterMatched {
-						continue
-					}
-				}
-				// Check asset fingerprint filter
-				if len(c.filterAssetFingerprints) > 0 {
-					filterMatched := false
-					for _, filterAssetFingerprint := range c.filterAssetFingerprints {
-						foundMatch := false
-						// Include resolved inputs as outputs for matching
-						allOutputs := append(v.Outputs, v.ResolvedInputs...)
-						for _, output := range allOutputs {
-							if output.Assets() != nil {
-								for _, policyId := range output.Assets().Policies() {
-									for _, assetName := range output.Assets().Assets(policyId) {
-										assetFp := ledger.NewAssetFingerprint(policyId.Bytes(), assetName)
-										if assetFp.String() == filterAssetFingerprint {
+					// Check policy ID filter
+					if len(c.filterPolicyIds) > 0 {
+						filterMatched := false
+						for _, filterPolicyId := range c.filterPolicyIds {
+							foundMatch := false
+							// Include resolved inputs as outputs for matching
+							allOutputs := append(v.Outputs, v.ResolvedInputs...)
+							for _, output := range allOutputs {
+								if output.Assets() != nil {
+									for _, policyId := range output.Assets().Policies() {
+										if policyId.String() == filterPolicyId {
 											foundMatch = true
+											break
+										}
+									}
+								}
+								if foundMatch {
+									break
+								}
+							}
+							if foundMatch {
+								filterMatched = true
+								break
+							}
+						}
+						// Skip the event if none of the filter values matched
+						if !filterMatched {
+							continue
+						}
+					}
+					// Check asset fingerprint filter
+					if len(c.filterAssetFingerprints) > 0 {
+						filterMatched := false
+						for _, filterAssetFingerprint := range c.filterAssetFingerprints {
+							foundMatch := false
+							// Include resolved inputs as outputs for matching
+							allOutputs := append(v.Outputs, v.ResolvedInputs...)
+							for _, output := range allOutputs {
+								if output.Assets() != nil {
+									for _, policyId := range output.Assets().Policies() {
+										for _, assetName := range output.Assets().Assets(policyId) {
+											assetFp := ledger.NewAssetFingerprint(policyId.Bytes(), assetName)
+											if assetFp.String() == filterAssetFingerprint {
+												foundMatch = true
+											}
+										}
+										if foundMatch {
+											break
 										}
 									}
 									if foundMatch {
 										break
 									}
 								}
-								if foundMatch {
-									break
-								}
+							}
+							if foundMatch {
+								filterMatched = true
+								break
 							}
 						}
-						if foundMatch {
-							filterMatched = true
-							break
+						// Skip the event if none of the filter values matched
+						if !filterMatched {
+							continue
 						}
 					}
-					// Skip the event if none of the filter values matched
-					if !filterMatched {
-						continue
+					// Check pool filter
+					if len(c.filterPoolIds) > 0 {
+						filterMatched := false
+						for _, filterPoolId := range c.filterPoolIds {
+							if filterMatched {
+								break
+							}
+							isPoolBech32 := strings.HasPrefix(filterPoolId, "pool")
+							foundMatch := false
+							for _, certificate := range v.Certificates {
+								switch cert := certificate.(type) {
+								case *ledger.StakeDelegationCertificate:
+									b := &ledger.Blake2b224{}
+									copy(b[:], cert.PoolKeyHash[:])
+									if b.String() == filterPoolId {
+										foundMatch = true
+									} else if isPoolBech32 {
+										// lifted from gouroboros/ledger
+										convData, err := bech32.ConvertBits(certificate.Cbor(), 8, 5, true)
+										if err != nil {
+											continue
+										}
+										encoded, err := bech32.Encode("pool", convData)
+										if err != nil {
+											continue
+										}
+										if encoded == filterPoolId {
+											foundMatch = true
+										}
+									}
+									if foundMatch {
+										filterMatched = true
+										break
+									}
+								case *ledger.PoolRetirementCertificate:
+									b := &ledger.Blake2b224{}
+									copy(b[:], cert.PoolKeyHash[:])
+									if b.String() == filterPoolId {
+										foundMatch = true
+									} else if isPoolBech32 {
+										// lifted from gouroboros/ledger
+										convData, err := bech32.ConvertBits(certificate.Cbor(), 8, 5, true)
+										if err != nil {
+											continue
+										}
+										encoded, err := bech32.Encode("pool", convData)
+										if err != nil {
+											continue
+										}
+										if encoded == filterPoolId {
+											foundMatch = true
+										}
+									}
+									if foundMatch {
+										filterMatched = true
+										break
+									}
+								case *ledger.PoolRegistrationCertificate:
+									b := &ledger.Blake2b224{}
+									copy(b[:], cert.Operator[:])
+									if b.String() == filterPoolId {
+										foundMatch = true
+									} else if isPoolBech32 {
+										// lifted from gouroboros/ledger
+										convData, err := bech32.ConvertBits(certificate.Cbor(), 8, 5, true)
+										if err != nil {
+											continue
+										}
+										encoded, err := bech32.Encode("pool", convData)
+										if err != nil {
+											continue
+										}
+										if encoded == filterPoolId {
+											foundMatch = true
+										}
+									}
+									if foundMatch {
+										filterMatched = true
+										break
+									}
+								}
+							}
+							if foundMatch {
+								filterMatched = true
+								break
+							}
+						}
+						// Skip the event if none of the filter values matched
+						if !filterMatched {
+							continue
+						}
 					}
 				}
-				// Check pool filter
-				if len(c.filterPoolIds) > 0 {
-					filterMatched := false
-					for _, filterPoolId := range c.filterPoolIds {
-						if filterMatched {
-							break
-						}
-						isPoolBech32 := strings.HasPrefix(filterPoolId, "pool")
-						foundMatch := false
-						for _, certificate := range v.Certificates {
-							switch cert := certificate.(type) {
-							case *ledger.StakeDelegationCertificate:
-								b := &ledger.Blake2b224{}
-								copy(b[:], cert.PoolKeyHash[:])
-								if b.String() == filterPoolId {
-									foundMatch = true
-								} else if isPoolBech32 {
-									// lifted from gouroboros/ledger
-									convData, err := bech32.ConvertBits(certificate.Cbor(), 8, 5, true)
-									if err != nil {
-										continue
-									}
-									encoded, err := bech32.Encode("pool", convData)
-									if err != nil {
-										continue
-									}
-									if encoded == filterPoolId {
-										foundMatch = true
-									}
-								}
-								if foundMatch {
-									filterMatched = true
-									break
-								}
-							case *ledger.PoolRetirementCertificate:
-								b := &ledger.Blake2b224{}
-								copy(b[:], cert.PoolKeyHash[:])
-								if b.String() == filterPoolId {
-									foundMatch = true
-								} else if isPoolBech32 {
-									// lifted from gouroboros/ledger
-									convData, err := bech32.ConvertBits(certificate.Cbor(), 8, 5, true)
-									if err != nil {
-										continue
-									}
-									encoded, err := bech32.Encode("pool", convData)
-									if err != nil {
-										continue
-									}
-									if encoded == filterPoolId {
-										foundMatch = true
-									}
-								}
-								if foundMatch {
-									filterMatched = true
-									break
-								}
-							case *ledger.PoolRegistrationCertificate:
-								b := &ledger.Blake2b224{}
-								copy(b[:], cert.Operator[:])
-								if b.String() == filterPoolId {
-									foundMatch = true
-								} else if isPoolBech32 {
-									// lifted from gouroboros/ledger
-									convData, err := bech32.ConvertBits(certificate.Cbor(), 8, 5, true)
-									if err != nil {
-										continue
-									}
-									encoded, err := bech32.Encode("pool", convData)
-									if err != nil {
-										continue
-									}
-									if encoded == filterPoolId {
-										foundMatch = true
-									}
-								}
-								if foundMatch {
-									filterMatched = true
-									break
-								}
-							}
-						}
-						if foundMatch {
-							filterMatched = true
-							break
-						}
-					}
-					// Skip the event if none of the filter values matched
-					if !filterMatched {
-						continue
-					}
+				// Send event along, but check for shutdown
+				select {
+				case <-c.doneChan:
+					return
+				case c.outputChan <- evt:
 				}
 			}
-			c.outputChan <- evt
 		}
 	}()
 	return nil
@@ -329,14 +343,26 @@ func (c *ChainSync) Start() error {
 
 // Stop the chain sync filter
 func (c *ChainSync) Stop() error {
-	close(c.inputChan)
-	close(c.outputChan)
+	c.stopOnce.Do(func() {
+		if c.doneChan != nil {
+			close(c.doneChan)
+		}
+		if c.inputChan != nil {
+			close(c.inputChan)
+		}
+		if c.outputChan != nil {
+			close(c.outputChan)
+		}
+		if c.errorChan != nil {
+			close(c.errorChan)
+		}
+	})
 	return nil
 }
 
-// SetErrorChan sets the error channel
-func (c *ChainSync) SetErrorChan(ch chan<- error) {
-	c.errorChan = ch
+// ErrorChan returns the plugin's error channel
+func (c *ChainSync) ErrorChan() <-chan error {
+	return c.errorChan
 }
 
 // InputChan returns the input event channel

--- a/filter/chainsync/chainsync_test.go
+++ b/filter/chainsync/chainsync_test.go
@@ -158,30 +158,36 @@ func TestChainSync_Stop(t *testing.T) {
 	if err != nil {
 		t.Fatalf("expected no error, got %v", err)
 	}
-	// Check if channels are closed
-	select {
-	case <-c.inputChan:
-	default:
-		t.Fatalf("expected inputChan to be closed")
+	// Check if channels are nil after stop
+	if c.inputChan != nil {
+		t.Fatalf("expected inputChan to be nil after stop")
 	}
-	select {
-	case <-c.outputChan:
-	default:
-		t.Fatalf("expected outputChan to be closed")
+	if c.outputChan != nil {
+		t.Fatalf("expected outputChan to be nil after stop")
 	}
 }
 
 func TestChainSync_InputChan(t *testing.T) {
 	c := New()
+	err := c.Start()
+	if err != nil {
+		t.Fatalf("expected no error, got %v", err)
+	}
+	defer c.Stop()
 	if c.InputChan() == nil {
-		t.Fatalf("expected non-nil inputChan")
+		t.Fatalf("expected non-nil inputChan after Start()")
 	}
 }
 
 func TestChainSync_OutputChan(t *testing.T) {
 	c := New()
+	err := c.Start()
+	if err != nil {
+		t.Fatalf("expected no error, got %v", err)
+	}
+	defer c.Stop()
 	if c.OutputChan() == nil {
-		t.Fatalf("expected non-nil outputChan")
+		t.Fatalf("expected non-nil outputChan after Start()")
 	}
 }
 

--- a/filter/chainsync/plugin_test.go
+++ b/filter/chainsync/plugin_test.go
@@ -50,11 +50,24 @@ func TestPluginChannels(t *testing.T) {
 	// Create a new plugin instance
 	p := NewFromCmdlineOptions()
 
+	// Start the plugin (channels are created in Start())
+	err := p.Start()
+	assert.NoError(t, err, "Plugin should start without errors")
+	defer p.Stop()
+
 	// Verify that the input channel is not nil
-	assert.NotNil(t, p.InputChan(), "Input channel should not be nil")
+	assert.NotNil(
+		t,
+		p.InputChan(),
+		"Input channel should not be nil after Start()",
+	)
 
 	// Verify that the output channel is not nil
-	assert.NotNil(t, p.OutputChan(), "Output channel should not be nil")
+	assert.NotNil(
+		t,
+		p.OutputChan(),
+		"Output channel should not be nil after Start()",
+	)
 }
 
 func TestPluginEventProcessing(t *testing.T) {

--- a/filter/event/event.go
+++ b/filter/event/event.go
@@ -16,25 +16,26 @@ package event
 
 import (
 	"slices"
+	"sync"
 
 	"github.com/blinklabs-io/adder/event"
 	"github.com/blinklabs-io/adder/plugin"
 )
 
 type Event struct {
-	errorChan   chan<- error
+	errorChan   chan error
 	inputChan   chan event.Event
 	outputChan  chan event.Event
+	doneChan    chan struct{}
+	wg          sync.WaitGroup
+	stopOnce    sync.Once
 	logger      plugin.Logger
 	filterTypes []string
 }
 
 // New returns a new Event object with the specified options applied
 func New(options ...EventOptionFunc) *Event {
-	e := &Event{
-		inputChan:  make(chan event.Event, 10),
-		outputChan: make(chan event.Event, 10),
-	}
+	e := &Event{}
 	for _, option := range options {
 		option(e)
 	}
@@ -43,22 +44,42 @@ func New(options ...EventOptionFunc) *Event {
 
 // Start the event filter
 func (e *Event) Start() error {
+	// Guard against double-start: wait for existing goroutine to exit
+	if e.doneChan != nil {
+		close(e.doneChan)
+		e.wg.Wait()
+	}
+	e.errorChan = make(chan error)
+	e.inputChan = make(chan event.Event, 10)
+	e.outputChan = make(chan event.Event, 10)
+	e.doneChan = make(chan struct{})
+	e.stopOnce = sync.Once{}
+	e.wg.Add(1)
 	go func() {
+		defer e.wg.Done()
 		for {
-			evt, ok := <-e.inputChan
-			// Channel has been closed, which means we're shutting down
-			if !ok {
+			select {
+			case <-e.doneChan:
 				return
-			}
-			// Drop events if we have a type filter configured and the event doesn't match
-			if len(e.filterTypes) > 0 {
-				matched := slices.Contains(e.filterTypes, evt.Type)
-				if !matched {
-					continue
+			case evt, ok := <-e.inputChan:
+				// Channel has been closed, which means we're shutting down
+				if !ok {
+					return
+				}
+				// Drop events if we have a type filter configured and the event doesn't match
+				if len(e.filterTypes) > 0 {
+					matched := slices.Contains(e.filterTypes, evt.Type)
+					if !matched {
+						continue
+					}
+				}
+				// Send event along, but check for shutdown
+				select {
+				case <-e.doneChan:
+					return
+				case e.outputChan <- evt:
 				}
 			}
-			// Send event along
-			e.outputChan <- evt
 		}
 	}()
 	return nil
@@ -66,14 +87,28 @@ func (e *Event) Start() error {
 
 // Stop the event filter
 func (e *Event) Stop() error {
-	close(e.inputChan)
-	close(e.outputChan)
+	e.stopOnce.Do(func() {
+		if e.doneChan != nil {
+			close(e.doneChan)
+		}
+		// Wait for goroutine to exit before closing channels
+		e.wg.Wait()
+		if e.inputChan != nil {
+			close(e.inputChan)
+		}
+		if e.outputChan != nil {
+			close(e.outputChan)
+		}
+		if e.errorChan != nil {
+			close(e.errorChan)
+		}
+	})
 	return nil
 }
 
-// SetErrorChan sets the error channel
-func (e *Event) SetErrorChan(ch chan<- error) {
-	e.errorChan = ch
+// ErrorChan returns the plugin's error channel
+func (e *Event) ErrorChan() <-chan error {
+	return e.errorChan
 }
 
 // InputChan returns the input event channel

--- a/input/chainsync/chainsync.go
+++ b/input/chainsync/chainsync.go
@@ -25,6 +25,7 @@ import (
 	"net/url"
 	"slices"
 	"strings"
+	"sync"
 	"time"
 
 	"github.com/SundaeSwap-finance/kugo"
@@ -88,10 +89,12 @@ type ChainSync struct {
 	logger             plugin.Logger
 	statusUpdateFunc   StatusUpdateFunc
 	blockfetchDoneChan chan struct{}
+	doneChan           chan struct{}
+	wg                 sync.WaitGroup
 	kupoClient         *kugo.Client
 	oConn              *ouroboros.Connection
 	eventChan          chan event.Event
-	errorChan          chan<- error
+	errorChan          chan error
 	status             *ChainSyncStatus
 	dialFamily         string
 	kupoUrl            string
@@ -128,7 +131,6 @@ type StatusUpdateFunc func(ChainSyncStatus)
 // New returns a new ChainSync object with the specified options applied
 func New(options ...ChainSyncOptionFunc) *ChainSync {
 	c := &ChainSync{
-		eventChan:       make(chan event.Event, 10),
 		intersectPoints: []ocommon.Point{},
 		status:          &ChainSyncStatus{},
 	}
@@ -142,6 +144,14 @@ func New(options ...ChainSyncOptionFunc) *ChainSync {
 
 // Start the chain sync input
 func (c *ChainSync) Start() error {
+	// Guard against double-start: signal existing goroutines to stop and wait
+	if c.doneChan != nil {
+		close(c.doneChan)
+		c.wg.Wait()
+	}
+	c.eventChan = make(chan event.Event, 10)
+	c.errorChan = make(chan error)
+	c.doneChan = make(chan struct{})
 	if err := c.setupConnection(); err != nil {
 		return err
 	}
@@ -166,14 +176,33 @@ func (c *ChainSync) Start() error {
 
 // Stop the chain sync input
 func (c *ChainSync) Stop() error {
-	err := c.oConn.Close()
-	close(c.eventChan)
+	var err error
+	// Signal goroutines to stop first
+	if c.doneChan != nil {
+		close(c.doneChan)
+		c.doneChan = nil
+	}
+	// Close connection (this unblocks the error handler goroutine)
+	if c.oConn != nil {
+		err = c.oConn.Close()
+		c.oConn = nil
+	}
+	// Wait for goroutines to exit before closing channels
+	c.wg.Wait()
+	if c.eventChan != nil {
+		close(c.eventChan)
+		c.eventChan = nil
+	}
+	if c.errorChan != nil {
+		close(c.errorChan)
+		c.errorChan = nil
+	}
 	return err
 }
 
-// SetErrorChan sets the error channel
-func (c *ChainSync) SetErrorChan(ch chan<- error) {
-	c.errorChan = ch
+// ErrorChan returns the plugin's error channel
+func (c *ChainSync) ErrorChan() <-chan error {
+	return c.errorChan
 }
 
 // InputChan always returns nil
@@ -257,6 +286,7 @@ func (c *ChainSync) setupConnection() error {
 		c.logger.Info("connected to node at " + c.dialAddress)
 	}
 	// Start async error handler
+	c.wg.Add(1)
 	go func() {
 		err, ok := <-c.oConn.ErrorChan()
 		if ok {
@@ -298,6 +328,10 @@ func (c *ChainSync) setupConnection() error {
 					if len(c.cursorCache) > 0 {
 						c.intersectPoints = c.cursorCache[:]
 					}
+					// Decrement WaitGroup before calling Start() to avoid deadlock.
+					// Start() calls wg.Wait(), so this goroutine must release its
+					// count first. The new Start() will spawn a fresh error handler.
+					c.wg.Done()
 					// Restart the connection
 					if err := c.Start(); err != nil {
 						if c.logger != nil {
@@ -307,22 +341,38 @@ func (c *ChainSync) setupConnection() error {
 								err,
 							))
 						}
+						// Re-increment since we need to try again in this goroutine
+						c.wg.Add(1)
 						continue
 					}
-					break
+					// Successfully restarted - this goroutine exits,
+					// new goroutine from Start() takes over
+					return
 				}
 			} else {
-				// Pass error through our own error channel
-				if c.errorChan != nil {
-					c.errorChan <- err
-				} else if c.logger != nil {
-					c.logger.Warn(fmt.Sprintf(
-						"error occurred but no error channel set: %s",
-						err,
-					))
+				// Pass error through our own error channel, but check for shutdown
+				select {
+				case <-c.doneChan:
+					c.wg.Done()
+					return
+				default:
+					if c.errorChan != nil {
+						select {
+						case <-c.doneChan:
+							c.wg.Done()
+							return
+						case c.errorChan <- err:
+						}
+					} else if c.logger != nil {
+						c.logger.Warn(fmt.Sprintf(
+							"error occurred but no error channel set: %s",
+							err,
+						))
+					}
 				}
 			}
 		}
+		c.wg.Done()
 	}()
 	return nil
 }

--- a/output/embedded/embedded.go
+++ b/output/embedded/embedded.go
@@ -16,6 +16,7 @@ package embedded
 
 import (
 	"fmt"
+	"sync"
 
 	"github.com/blinklabs-io/adder/event"
 )
@@ -23,16 +24,16 @@ import (
 type CallbackFunc func(event.Event) error
 
 type EmbeddedOutput struct {
-	errorChan    chan<- error
+	errorChan    chan error
 	eventChan    chan event.Event
+	doneChan     chan struct{}
+	wg           sync.WaitGroup
 	callbackFunc CallbackFunc
 	outputChan   chan event.Event
 }
 
 func New(options ...EmbeddedOptionFunc) *EmbeddedOutput {
-	e := &EmbeddedOutput{
-		eventChan: make(chan event.Event, 10),
-	}
+	e := &EmbeddedOutput{}
 	for _, option := range options {
 		option(e)
 	}
@@ -41,23 +42,44 @@ func New(options ...EmbeddedOptionFunc) *EmbeddedOutput {
 
 // Start the embedded output
 func (e *EmbeddedOutput) Start() error {
+	// Guard against double-start: wait for existing goroutine to exit
+	if e.doneChan != nil {
+		close(e.doneChan)
+		e.wg.Wait()
+	}
+	e.eventChan = make(chan event.Event, 10)
+	e.errorChan = make(chan error)
+	e.doneChan = make(chan struct{})
+	e.wg.Add(1)
 	go func() {
+		defer e.wg.Done()
 		for {
-			evt, ok := <-e.eventChan
-			// Channel has been closed, which means we're shutting down
-			if !ok {
+			select {
+			case <-e.doneChan:
 				return
-			}
-			if e.callbackFunc != nil {
-				if err := e.callbackFunc(evt); err != nil {
-					if e.errorChan != nil {
-						e.errorChan <- fmt.Errorf("callback function error: %w", err)
-					}
+			case evt, ok := <-e.eventChan:
+				// Channel has been closed, which means we're shutting down
+				if !ok {
 					return
 				}
-			}
-			if e.outputChan != nil {
-				e.outputChan <- evt
+				if e.callbackFunc != nil {
+					if err := e.callbackFunc(evt); err != nil {
+						// Use select to safely send error without racing with Stop()
+						select {
+						case <-e.doneChan:
+							return
+						case e.errorChan <- fmt.Errorf("callback function error: %w", err):
+						}
+						return
+					}
+				}
+				if e.outputChan != nil {
+					select {
+					case <-e.doneChan:
+						return
+					case e.outputChan <- evt:
+					}
+				}
 			}
 		}
 	}()
@@ -66,16 +88,30 @@ func (e *EmbeddedOutput) Start() error {
 
 // Stop the embedded output
 func (e *EmbeddedOutput) Stop() error {
-	close(e.eventChan)
+	if e.doneChan != nil {
+		close(e.doneChan)
+		e.doneChan = nil
+	}
+	// Wait for goroutine to exit before closing channels
+	e.wg.Wait()
+	if e.eventChan != nil {
+		close(e.eventChan)
+		e.eventChan = nil
+	}
+	if e.errorChan != nil {
+		close(e.errorChan)
+		e.errorChan = nil
+	}
 	if e.outputChan != nil {
 		close(e.outputChan)
+		e.outputChan = nil
 	}
 	return nil
 }
 
-// SetErrorChan sets the error channel
-func (e *EmbeddedOutput) SetErrorChan(ch chan<- error) {
-	e.errorChan = ch
+// ErrorChan returns the plugin's error channel
+func (e *EmbeddedOutput) ErrorChan() <-chan error {
+	return e.errorChan
 }
 
 // InputChan returns the input event channel

--- a/output/log/log.go
+++ b/output/log/log.go
@@ -24,7 +24,7 @@ import (
 )
 
 type LogOutput struct {
-	errorChan    chan<- error
+	errorChan    chan error
 	eventChan    chan event.Event
 	logger       plugin.Logger
 	outputLogger *slog.Logger
@@ -33,8 +33,7 @@ type LogOutput struct {
 
 func New(options ...LogOptionFunc) *LogOutput {
 	l := &LogOutput{
-		eventChan: make(chan event.Event, 10),
-		level:     "info",
+		level: "info",
 	}
 	for _, option := range options {
 		option(l)
@@ -54,6 +53,8 @@ func New(options ...LogOptionFunc) *LogOutput {
 
 // Start the log output
 func (l *LogOutput) Start() error {
+	l.eventChan = make(chan event.Event, 10)
+	l.errorChan = make(chan error)
 	go func() {
 		for {
 			evt, ok := <-l.eventChan
@@ -79,13 +80,20 @@ func (l *LogOutput) Start() error {
 
 // Stop the log output
 func (l *LogOutput) Stop() error {
-	close(l.eventChan)
+	if l.eventChan != nil {
+		close(l.eventChan)
+		l.eventChan = nil
+	}
+	if l.errorChan != nil {
+		close(l.errorChan)
+		l.errorChan = nil
+	}
 	return nil
 }
 
-// SetErrorChan sets the error channel
-func (l *LogOutput) SetErrorChan(ch chan<- error) {
-	l.errorChan = ch
+// ErrorChan returns the plugin's error channel
+func (l *LogOutput) ErrorChan() <-chan error {
+	return l.errorChan
 }
 
 // InputChan returns the input event channel

--- a/output/notify/notify.go
+++ b/output/notify/notify.go
@@ -28,7 +28,7 @@ import (
 var icon []byte
 
 type NotifyOutput struct {
-	errorChan chan<- error
+	errorChan chan error
 	eventChan chan event.Event
 	logger    plugin.Logger
 	title     string
@@ -36,8 +36,7 @@ type NotifyOutput struct {
 
 func New(options ...NotifyOptionFunc) *NotifyOutput {
 	n := &NotifyOutput{
-		eventChan: make(chan event.Event, 10),
-		title:     "Adder",
+		title: "Adder",
 	}
 	for _, option := range options {
 		option(n)
@@ -47,6 +46,8 @@ func New(options ...NotifyOptionFunc) *NotifyOutput {
 
 // Start the notify output
 func (n *NotifyOutput) Start() error {
+	n.eventChan = make(chan event.Event, 10)
+	n.errorChan = make(chan error)
 	// Write our icon asset
 	userCacheDir, err := os.UserCacheDir()
 	if err != nil {
@@ -164,13 +165,20 @@ func (n *NotifyOutput) Start() error {
 
 // Stop the embedded output
 func (n *NotifyOutput) Stop() error {
-	close(n.eventChan)
+	if n.eventChan != nil {
+		close(n.eventChan)
+		n.eventChan = nil
+	}
+	if n.errorChan != nil {
+		close(n.errorChan)
+		n.errorChan = nil
+	}
 	return nil
 }
 
-// SetErrorChan sets the error channel
-func (n *NotifyOutput) SetErrorChan(ch chan<- error) {
-	n.errorChan = ch
+// ErrorChan returns the plugin's error channel
+func (n *NotifyOutput) ErrorChan() <-chan error {
+	return n.errorChan
 }
 
 // InputChan returns the input event channel

--- a/output/push/push.go
+++ b/output/push/push.go
@@ -32,7 +32,7 @@ import (
 )
 
 type PushOutput struct {
-	errorChan              chan<- error
+	errorChan              chan error
 	eventChan              chan event.Event
 	logger                 plugin.Logger
 	accessToken            string
@@ -53,9 +53,7 @@ type PushPayload struct {
 }
 
 func New(options ...PushOptionFunc) *PushOutput {
-	p := &PushOutput{
-		eventChan: make(chan event.Event, 10),
-	}
+	p := &PushOutput{}
 	for _, option := range options {
 		option(p)
 	}
@@ -69,6 +67,8 @@ func New(options ...PushOptionFunc) *PushOutput {
 }
 
 func (p *PushOutput) Start() error {
+	p.eventChan = make(chan event.Event, 10)
+	p.errorChan = make(chan error)
 	logger := logging.GetLogger()
 	logger.Info("starting push notification server")
 	go func() {
@@ -284,13 +284,20 @@ func (p *PushOutput) GetProjectId() error {
 
 // Stop the embedded output
 func (p *PushOutput) Stop() error {
-	close(p.eventChan)
+	if p.eventChan != nil {
+		close(p.eventChan)
+		p.eventChan = nil
+	}
+	if p.errorChan != nil {
+		close(p.errorChan)
+		p.errorChan = nil
+	}
 	return nil
 }
 
-// SetErrorChan sets the error channel
-func (p *PushOutput) SetErrorChan(ch chan<- error) {
-	p.errorChan = ch
+// ErrorChan returns the plugin's error channel
+func (p *PushOutput) ErrorChan() <-chan error {
+	return p.errorChan
 }
 
 // InputChan returns the input event channel

--- a/output/webhook/webhook.go
+++ b/output/webhook/webhook.go
@@ -24,6 +24,7 @@ import (
 	"io"
 	"net/http"
 	"strconv"
+	"sync"
 	"time"
 
 	"github.com/blinklabs-io/adder/event"
@@ -39,8 +40,10 @@ const (
 )
 
 type WebhookOutput struct {
-	errorChan  chan<- error
+	errorChan  chan error
 	eventChan  chan event.Event
+	doneChan   chan struct{}
+	wg         sync.WaitGroup
 	logger     plugin.Logger
 	format     string
 	url        string
@@ -51,7 +54,6 @@ type WebhookOutput struct {
 
 func New(options ...WebhookOptionFunc) *WebhookOutput {
 	w := &WebhookOutput{
-		eventChan:  make(chan event.Event, 10),
 		format:     "adder",
 		url:        "http://localhost:3000",
 		skipVerify: false,
@@ -64,43 +66,57 @@ func New(options ...WebhookOptionFunc) *WebhookOutput {
 
 // Start the webhook output
 func (w *WebhookOutput) Start() error {
+	// Guard against double-start: wait for existing goroutine to exit
+	if w.doneChan != nil {
+		close(w.doneChan)
+		w.wg.Wait()
+	}
+	w.eventChan = make(chan event.Event, 10)
+	w.errorChan = make(chan error)
+	w.doneChan = make(chan struct{})
 	logger := logging.GetLogger()
 	logger.Info("starting webhook server")
+	w.wg.Add(1)
 	go func() {
+		defer w.wg.Done()
 		for {
-			evt, ok := <-w.eventChan
-			// Channel has been closed, which means we're shutting down
-			if !ok {
+			select {
+			case <-w.doneChan:
 				return
-			}
-			payload := evt.Payload
-			if payload == nil {
-				panic(fmt.Errorf("ERROR: %v", payload))
-			}
-			context := evt.Context
-			switch evt.Type {
-			case "chainsync.block":
-				if context == nil {
-					panic(fmt.Errorf("ERROR: %v", context))
+			case evt, ok := <-w.eventChan:
+				// Channel has been closed, which means we're shutting down
+				if !ok {
+					return
 				}
-				be := payload.(event.BlockEvent)
-				bc := context.(event.BlockContext)
-				evt.Payload = be
-				evt.Context = bc
-			case "chainsync.rollback":
-				re := payload.(event.RollbackEvent)
-				evt.Payload = re
-			case "chainsync.transaction":
-				te := payload.(event.TransactionEvent)
-				evt.Payload = te
-			default:
-				logger.Error("unknown event type: " + evt.Type)
-				return
-			}
-			// TODO: error handle (#334)
-			err := w.SendWebhook(&evt)
-			if err != nil {
-				logger.Error(fmt.Sprintf("ERROR: %s", err))
+				payload := evt.Payload
+				if payload == nil {
+					panic(fmt.Errorf("ERROR: %v", payload))
+				}
+				context := evt.Context
+				switch evt.Type {
+				case "chainsync.block":
+					if context == nil {
+						panic(fmt.Errorf("ERROR: %v", context))
+					}
+					be := payload.(event.BlockEvent)
+					bc := context.(event.BlockContext)
+					evt.Payload = be
+					evt.Context = bc
+				case "chainsync.rollback":
+					re := payload.(event.RollbackEvent)
+					evt.Payload = re
+				case "chainsync.transaction":
+					te := payload.(event.TransactionEvent)
+					evt.Payload = te
+				default:
+					logger.Error("unknown event type: " + evt.Type)
+					return
+				}
+				// TODO: error handle (#334)
+				err := w.SendWebhook(&evt)
+				if err != nil {
+					logger.Error(fmt.Sprintf("ERROR: %s", err))
+				}
 			}
 		}
 	}()
@@ -299,15 +315,28 @@ func (w *WebhookOutput) SendWebhook(e *event.Event) error {
 	return nil
 }
 
-// Stop the embedded output
+// Stop the webhook output
 func (w *WebhookOutput) Stop() error {
-	close(w.eventChan)
+	if w.doneChan != nil {
+		close(w.doneChan)
+		w.doneChan = nil
+	}
+	// Wait for goroutine to exit before closing channels
+	w.wg.Wait()
+	if w.eventChan != nil {
+		close(w.eventChan)
+		w.eventChan = nil
+	}
+	if w.errorChan != nil {
+		close(w.errorChan)
+		w.errorChan = nil
+	}
 	return nil
 }
 
-// SetErrorChan sets the error channel
-func (w *WebhookOutput) SetErrorChan(ch chan<- error) {
-	w.errorChan = ch
+// ErrorChan returns the plugin's error channel
+func (w *WebhookOutput) ErrorChan() <-chan error {
+	return w.errorChan
 }
 
 // InputChan returns the input event channel

--- a/pipeline/pipeline.go
+++ b/pipeline/pipeline.go
@@ -24,18 +24,15 @@ import (
 )
 
 type Pipeline struct {
-	filterChan       chan event.Event
-	outputChan       chan event.Event
-	errorChan        chan error
-	doneChan         chan bool
-	inputs           []plugin.Plugin
-	filters          []plugin.Plugin
-	outputs          []plugin.Plugin
-	inputErrorChans  []chan error
-	filterErrorChans []chan error
-	outputErrorChans []chan error
-	wg               sync.WaitGroup
-	stopOnce         sync.Once
+	filterChan chan event.Event
+	outputChan chan event.Event
+	errorChan  chan error
+	doneChan   chan bool
+	inputs     []plugin.Plugin
+	filters    []plugin.Plugin
+	outputs    []plugin.Plugin
+	wg         sync.WaitGroup
+	stopOnce   sync.Once
 }
 
 func New() *Pipeline {
@@ -60,28 +57,33 @@ func (p *Pipeline) AddOutput(output plugin.Plugin) {
 	p.outputs = append(p.outputs, output)
 }
 
-// ErrorChan is read-only
+// ErrorChan returns the pipeline's error channel (read-only).
+// Note: After calling Stop() and Start() to restart the pipeline,
+// consumers must call ErrorChan() again to get the new channel.
+// References obtained before restart will point to a closed channel.
 func (p *Pipeline) ErrorChan() <-chan error {
 	return p.errorChan
 }
 
-// Start initiates the configured plugins and starts the necessary background processes to run the pipeline
+// Start initiates the configured plugins and starts the necessary background processes to run the pipeline.
+// A stopped pipeline can be restarted by calling Start() again.
+// Note: After restart, consumers must re-obtain channels via ErrorChan() as the old channels are closed.
 func (p *Pipeline) Start() error {
-	// Check if doneChan is already closed this happens if pipeline was stopped
-	// A stopped pipeline cannot be restarted
+	// Check if doneChan is already closed (pipeline was stopped)
+	// If so, recreate channels to allow restart
 	select {
 	case <-p.doneChan:
-		return errors.New("cannot start a stopped pipeline")
+		p.doneChan = make(chan bool)
+		p.filterChan = make(chan event.Event)
+		p.outputChan = make(chan event.Event)
+		p.errorChan = make(chan error)
+		p.stopOnce = sync.Once{}
 	default:
 		// continue
 	}
 
 	// Start inputs
 	for _, input := range p.inputs {
-		// Create error channel for this plugin
-		errorChan := make(chan error)
-		input.SetErrorChan(errorChan)
-		p.inputErrorChans = append(p.inputErrorChans, errorChan)
 		if err := input.Start(); err != nil {
 			return fmt.Errorf("failed to start input: %w", err)
 		}
@@ -90,14 +92,10 @@ func (p *Pipeline) Start() error {
 		go p.chanCopyLoop(input.OutputChan(), p.filterChan)
 		// Start background error listener
 		p.wg.Add(1)
-		go p.errorChanWait(errorChan)
+		go p.errorChanWait(input.ErrorChan())
 	}
 	// Start filters
 	for idx, filter := range p.filters {
-		// Create error channel for this plugin
-		errorChan := make(chan error)
-		filter.SetErrorChan(errorChan)
-		p.filterErrorChans = append(p.filterErrorChans, errorChan)
 		if err := filter.Start(); err != nil {
 			return fmt.Errorf("failed to start filter: %w", err)
 		}
@@ -117,7 +115,7 @@ func (p *Pipeline) Start() error {
 		}
 		// Start background error listener
 		p.wg.Add(1)
-		go p.errorChanWait(errorChan)
+		go p.errorChanWait(filter.ErrorChan())
 	}
 	if len(p.filters) == 0 {
 		// Start background process to send events from combined filter channel to combined output channel if
@@ -127,16 +125,12 @@ func (p *Pipeline) Start() error {
 	}
 	// Start outputs
 	for _, output := range p.outputs {
-		// Create error channel for this plugin
-		errorChan := make(chan error)
-		output.SetErrorChan(errorChan)
-		p.outputErrorChans = append(p.outputErrorChans, errorChan)
 		if err := output.Start(); err != nil {
 			return fmt.Errorf("failed to start output: %w", err)
 		}
 		// Start background error listener
 		p.wg.Add(1)
-		go p.errorChanWait(errorChan)
+		go p.errorChanWait(output.ErrorChan())
 	}
 	p.wg.Add(1)
 	go p.outputChanLoop()
@@ -145,7 +139,7 @@ func (p *Pipeline) Start() error {
 
 // Stop shuts down the pipeline and all plugins
 // Stop is idempotent and safe to call multiple times
-// A stopped pipeline cannot be restarted
+// A stopped pipeline can be restarted by calling Start() again
 func (p *Pipeline) Stop() error {
 	var stopErrors []error
 

--- a/pipeline/pipeline_test.go
+++ b/pipeline/pipeline_test.go
@@ -1,27 +1,49 @@
 package pipeline
 
 import (
+	"sync"
 	"testing"
+	"time"
 
 	"github.com/blinklabs-io/adder/event"
 )
 
 // plugin that panics when Stop is called
-type panicPlugin struct{}
+type panicPlugin struct {
+	errChanOnce sync.Once
+	errChan     chan error
+}
 
 func (p *panicPlugin) Start() error { return nil }
 func (p *panicPlugin) Stop() error  { panic("stop panic") }
 
-func (p *panicPlugin) SetErrorChan(chan<- error)      {}
+func (p *panicPlugin) ErrorChan() <-chan error {
+	p.errChanOnce.Do(func() {
+		if p.errChan == nil {
+			p.errChan = make(chan error)
+		}
+	})
+	return p.errChan
+}
 func (p *panicPlugin) InputChan() chan<- event.Event  { return nil }
 func (p *panicPlugin) OutputChan() <-chan event.Event { return nil }
 
 // simple no-op plugin
-type noopPlugin struct{}
+type noopPlugin struct {
+	errChanOnce sync.Once
+	errChan     chan error
+}
 
-func (n *noopPlugin) Start() error                   { return nil }
-func (n *noopPlugin) Stop() error                    { return nil }
-func (n *noopPlugin) SetErrorChan(chan<- error)      {}
+func (n *noopPlugin) Start() error { return nil }
+func (n *noopPlugin) Stop() error  { return nil }
+func (n *noopPlugin) ErrorChan() <-chan error {
+	n.errChanOnce.Do(func() {
+		if n.errChan == nil {
+			n.errChan = make(chan error)
+		}
+	})
+	return n.errChan
+}
 func (n *noopPlugin) InputChan() chan<- event.Event  { return nil }
 func (n *noopPlugin) OutputChan() <-chan event.Event { return nil }
 
@@ -50,5 +72,187 @@ func TestStopIdempotent(t *testing.T) {
 	}
 	if err := p.Stop(); err != nil {
 		t.Fatalf("unexpected error on second Stop (idempotent): %v", err)
+	}
+}
+
+func TestPipelineRestart(t *testing.T) {
+	p := New()
+	np := &noopPlugin{}
+	p.AddInput(np)
+
+	// First start
+	if err := p.Start(); err != nil {
+		t.Fatalf("unexpected error on first Start: %v", err)
+	}
+
+	// Stop
+	if err := p.Stop(); err != nil {
+		t.Fatalf("unexpected error on Stop: %v", err)
+	}
+
+	// Second start
+	if err := p.Start(); err != nil {
+		t.Fatalf("unexpected error on second Start: %v", err)
+	}
+
+	// Stop again
+	if err := p.Stop(); err != nil {
+		t.Fatalf("unexpected error on second Stop: %v", err)
+	}
+}
+
+// restartablePlugin is a plugin that properly supports restart by recreating channels
+type restartablePlugin struct {
+	errorChan  chan error
+	inputChan  chan event.Event
+	outputChan chan event.Event
+	doneChan   chan struct{}
+	stopOnce   sync.Once
+	wg         sync.WaitGroup
+	received   []event.Event
+	mu         sync.Mutex
+}
+
+func (r *restartablePlugin) Start() error {
+	r.mu.Lock()
+	r.errorChan = make(chan error)
+	r.inputChan = make(chan event.Event, 10)
+	r.outputChan = make(chan event.Event, 10)
+	r.doneChan = make(chan struct{})
+	r.stopOnce = sync.Once{}
+	r.received = nil
+	r.mu.Unlock()
+
+	r.wg.Add(1)
+	go func() {
+		defer r.wg.Done()
+		for {
+			select {
+			case <-r.doneChan:
+				return
+			case evt, ok := <-r.inputChan:
+				if !ok {
+					return
+				}
+				r.mu.Lock()
+				r.received = append(r.received, evt)
+				r.mu.Unlock()
+				select {
+				case r.outputChan <- evt:
+				case <-r.doneChan:
+					return
+				}
+			}
+		}
+	}()
+	return nil
+}
+
+func (r *restartablePlugin) Stop() error {
+	r.stopOnce.Do(func() {
+		if r.doneChan != nil {
+			close(r.doneChan)
+		}
+		// Wait for goroutine to exit before closing other channels
+		r.wg.Wait()
+		if r.inputChan != nil {
+			close(r.inputChan)
+		}
+		if r.outputChan != nil {
+			close(r.outputChan)
+		}
+		if r.errorChan != nil {
+			close(r.errorChan)
+		}
+	})
+	return nil
+}
+
+func (r *restartablePlugin) ErrorChan() <-chan error { return r.errorChan }
+
+func (r *restartablePlugin) InputChan() chan<- event.Event { return r.inputChan }
+
+func (r *restartablePlugin) OutputChan() <-chan event.Event { return r.outputChan }
+
+func (r *restartablePlugin) getReceived() []event.Event {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	result := make([]event.Event, len(r.received))
+	copy(result, r.received)
+	return result
+}
+
+// TestPipelineRestartWithEvents tests the full start -> process events -> stop -> start -> process events -> stop cycle
+func TestPipelineRestartWithEvents(t *testing.T) {
+	p := New()
+	input := &restartablePlugin{}
+	output := &restartablePlugin{}
+	p.AddInput(input)
+	p.AddOutput(output)
+
+	// First start
+	if err := p.Start(); err != nil {
+		t.Fatalf("unexpected error on first Start: %v", err)
+	}
+
+	// Send an event through the pipeline
+	evt1 := event.Event{Type: "test.event1"}
+	input.outputChan <- evt1
+
+	// Wait for event to be processed by the output plugin
+	deadline := time.Now().Add(time.Second)
+	for time.Now().Before(deadline) {
+		received := output.getReceived()
+		if len(received) > 0 && received[0].Type == evt1.Type {
+			break
+		}
+		time.Sleep(10 * time.Millisecond)
+	}
+
+	received := output.getReceived()
+	if len(received) != 1 || received[0].Type != evt1.Type {
+		t.Fatalf(
+			"expected 1 event with type %s after first start, got %d events",
+			evt1.Type,
+			len(received),
+		)
+	}
+
+	// Stop
+	if err := p.Stop(); err != nil {
+		t.Fatalf("unexpected error on Stop: %v", err)
+	}
+
+	// Second start (restart)
+	if err := p.Start(); err != nil {
+		t.Fatalf("unexpected error on second Start (restart): %v", err)
+	}
+
+	// Send another event through the pipeline after restart
+	evt2 := event.Event{Type: "test.event2"}
+	input.outputChan <- evt2
+
+	// Wait for event to be processed by the output plugin
+	deadline = time.Now().Add(time.Second)
+	for time.Now().Before(deadline) {
+		received := output.getReceived()
+		if len(received) > 0 && received[0].Type == evt2.Type {
+			break
+		}
+		time.Sleep(10 * time.Millisecond)
+	}
+
+	received = output.getReceived()
+	if len(received) != 1 || received[0].Type != evt2.Type {
+		t.Fatalf(
+			"expected 1 event with type %s after restart, got %d events",
+			evt2.Type,
+			len(received),
+		)
+	}
+
+	// Stop again
+	if err := p.Stop(); err != nil {
+		t.Fatalf("unexpected error on second Stop: %v", err)
 	}
 }

--- a/plugin/plugin.go
+++ b/plugin/plugin.go
@@ -19,12 +19,11 @@ import (
 )
 
 // Plugin represents a pluggable component in the pipeline.
-// The pipeline injects an error channel via SetErrorChan; plugins send errors but never close channels.
-// A nil error channel means "no error reporting"; plugins must handle this gracefully.
+// Plugins own their error channel; the pipeline reads from it.
 type Plugin interface {
 	Start() error
 	Stop() error
-	SetErrorChan(chan<- error)
+	ErrorChan() <-chan error
 	InputChan() chan<- event.Event
 	OutputChan() <-chan event.Event
 }

